hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkKafkaDataSetWriter.java

```diff
@@ -0,0 +1,138 @@
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License along with Splice Machine.
+ * If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.splicemachine.derby.stream.spark;
+
+import com.splicemachine.db.iapi.error.StandardException;
+import com.splicemachine.db.iapi.sql.execute.ExecRow;
+import com.splicemachine.db.iapi.types.SQLLongint;
+import com.splicemachine.db.impl.sql.execute.ValueRow;
+import com.splicemachine.derby.impl.SpliceSpark;
+import com.splicemachine.derby.stream.iapi.DataSet;
+import com.splicemachine.derby.stream.output.DataSetWriter;
+import com.splicemachine.si.api.txn.TxnView;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.util.LongAccumulator;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.Collections;
+import java.util.Optional;
+
+
+public class SparkKafkaDataSetWriter<V> implements DataSetWriter{
+    private String topicName;
+    private JavaRDD<V> rdd;
+    private Optional<StructType> schema;
+
+    public SparkKafkaDataSetWriter(){
+    }
+
+    public SparkKafkaDataSetWriter(JavaRDD<V> rdd,
+                                   String topicName){
+        this(rdd, topicName, Optional.empty());
+    }
+
+    public SparkKafkaDataSetWriter(JavaRDD<V> rdd,
+                                   String topicName,
+                                   Optional<StructType> schema){
+        this.rdd = rdd;
+        this.topicName = topicName;
+        this.schema = schema;
+    }
+
+    public void writeExternal(ObjectOutput out) throws IOException{
+        out.writeUTF(topicName);
+        out.writeObject(rdd);
+        out.writeObject(schema.orElse(new StructType()));
+    }
+
+    @SuppressWarnings("unchecked")
+    public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException{
+        topicName = in.readUTF();
+        rdd = (JavaRDD)in.readObject();
+        StructType structType = (StructType)in.readObject();
+        schema = structType.length() > 0 ? Optional.of(structType) : Optional.empty();
+    }
+
+    @Override
+    public DataSet<ExecRow> write() throws StandardException{
+        long start = System.currentTimeMillis();
+
+        CountFunction countFunction = new CountFunction<>();
+        KafkaStreamer kafkaStreamer = new KafkaStreamer(rdd.getNumPartitions(), topicName, schema);
+        JavaRDD streamed = rdd.map(countFunction).mapPartitionsWithIndex(kafkaStreamer, true);
+        streamed.collect();
+
+        Long count = countFunction.getCount().value();
+        if(count == 0) {
+            try {
+                kafkaStreamer.noData();
+            } catch(Exception e) {
```

Comment:
Why this is necessary?

Reply:
Previously, if a select statement legitimately returned no rows, the code block before this one would do nothing, and the client (KafkaToDF) would time out.  With this fix, the noData function sends an empty object with a schema so that the client is notified the query returned no rows and it also gets the schema. 